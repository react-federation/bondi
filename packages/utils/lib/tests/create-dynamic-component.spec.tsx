import React from 'react'
import { render, waitFor } from '@testing-library/react'
import createDynamicComponent from '../create-dynamic-component'
import useDynamicScript from '../use-dynamic-script'
import loadModule from '../load-module'

jest.mock('../use-dynamic-script')
jest.mock('../load-module')

describe('createDynamicComponent', () => {
  test('should show the placeholder and the loaded component', async () => {
    // @ts-ignore
    useDynamicScript.mockImplementationOnce(() => ({ ready: false }))
    // @ts-ignore
    loadModule.mockImplementationOnce(
      () => () =>
        new Promise((resolve) => {
          setTimeout(() => {
            resolve({ default: () => <p>test</p> })
          }, 300)
        })
    )
    const Component = createDynamicComponent({ url: 'https://localhost/script.js', scope: 'module', module: 'module' })
    const { asFragment } = render(<Component />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render the loaded compoenent', async () => {
    // @ts-ignore
    useDynamicScript.mockImplementationOnce(() => ({ ready: true }))
    // @ts-ignore
    loadModule.mockImplementationOnce(
      () => () =>
        new Promise((resolve) => {
          setTimeout(() => {
            resolve({ default: () => <p>test</p> })
          }, 300)
        })
    )
    const Component = createDynamicComponent({ url: 'https://localhost/script.js', scope: 'module', module: 'module' })
    const { getByText } = render(<Component />)

    await waitFor(() => expect(getByText('test')).toBeInTheDocument())
  })
})
