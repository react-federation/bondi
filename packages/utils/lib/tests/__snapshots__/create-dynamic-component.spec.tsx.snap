// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDynamicComponent should show the placeholder and the loaded component 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
